src/client.rs

```diff
@@ -0,0 +1,802 @@
+use jsonrpc::client::Client as RpcClient;
+use jsonrpc::error::Error;
+
+use crate::primitives::*;
+
+pub struct Client {
+    agent: RpcClient,
+}
+
+impl Client {
+    pub fn new(host: &str) -> Client {
+        Client {
+            agent: RpcClient::new(host.to_owned(), None, None),
+        }
+    }
+
+    pub fn new_with_credentials(host: &str, username: &str, password: &str) -> Client {
```

Comment:
```suggestion
    pub fn new_with_credentials(host: String, username: String, password: String) -> Client {
```
I would change the arguments to `String` here.
The reason is as follows:
- If you already have a `String` and want to call the function, you can pass it in without triggering a clone (what `to_owned()` does).
- If you only have a `&str`, it doesn't matter whether you perform the clone outside or inside the call.

Reply:
Of course also remove the `.to_owned()` when doing this change.